### PR TITLE
Basic states

### LINKED ISSUE #2: Scene Manager
**TODO**
- [ ] 

- Scene is View + GUI or only GUI if view is not needed

### PR DESCRIPTION
**Pull request by Kubaaa96**
Basic idea is to Change state when new is added to Stack stored in StateMachine.
Inspired by [Small Game Engine](https://github.com/eXpl0it3r/SmallGameEngine)
- Added StateMachine to manage States and deciding what to do with them.
- Changed Root widget to Group from Panel in SystemWindow 
- Every State need have Group to set Root by ```m_window.setRootGroup(*group);```
- Temporary DrawGUI function returning Root Group in State
- Removed m_currentState from System Window
- States are most of the time created inside another states by using ``` StateMachine::build<state>(machine, window, replace);```

### CODE REVIEW

File: src/client/states/GameState.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    struct GameState : core::state::State
+    {
+        GameState(core::state::StateMachine& machine, core::gui::SystemWindow& window, bool replace = true);
+
+        void pause() override;
+        void resume() override;
+
+        void draw() override;
+
+        core::gui::Group* drawGUI() override;
```
Comment (Sopel97) [inline src/client/states/GameState.h]: bad name, it doesn't draw it.

File: src/core/states/State.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+{
+    struct SystemWindow;
+}
+namespace ire::core::state
+{       
+    struct StateMachine;
+
+    struct State
+    {
+        State(StateMachine& machine, gui::SystemWindow& window, bool replace = true);
```
Comment (Sopel97) [inline src/core/states/State.h]: IMO replace=true is a bad default. I would leave it without a default value

File: src/core/states/StateMachine.h [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        bool running() const;
+        void quit();
+
+        template <typename T>
+        static std::unique_ptr<T> build(StateMachine& machine, gui::SystemWindow& window, bool replace = true);
+    private:
+        // The stack of states
+        std::stack<std::unique_ptr<State>> m_states;
+
+        bool m_resume;
```
Comment (Sopel97) [inline src/core/states/StateMachine.h]: what's the meaning of this? I can't see how the name relates to the functionality, and frankly I don't see the functionality at all too.

File: src/core/states/StateMachine.h [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        bool m_running;
+    };
+}
+
+template <typename T>
+std::unique_ptr<T> ire::core::state::StateMachine::build(ire::core::state::StateMachine& machine, ire::core::gui::SystemWindow& window, bool replace)
+{
+    return std::make_unique<T>(machine, window, replace);
+}
+
```
Comment (Sopel97) [inline src/core/states/StateMachine.h]: What's the purpose of this wrapper over make_unique? the name suggests it builds a state machine, but it just creates a state?

File: src/core/states/State.h [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        virtual void resume() = 0;
+
+        virtual void draw() = 0;
+
+        std::unique_ptr<State> next();
+
+        bool isReplacing() const;
+
+    protected:
+        StateMachine& m_machine;
```
Comment (Sopel97) [inline src/core/states/State.h]: m_machine is a very non-descriptive name for this. There can be many types of 'machines'

**Review by Sopel97 — CHANGES_REQUESTED** [review]

**Comment by Kubaaa96** [comment]
Example of current Working version
Example

File: src/core/states/StateMachine.h (hunk 4 above)
Comment (Kubaaa96) [reply src/core/states/StateMachine.h]: I removed it for now

**Review by Sopel97 — APPROVED** [review]
Looks good to me now.

**Pull request merged**